ports/esp32/modsocket.c

```diff
@@ -364,16 +366,103 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_1(socket_accept_obj, socket_accept);
 STATIC mp_obj_t socket_connect(const mp_obj_t arg0, const mp_obj_t arg1) {
     socket_obj_t *self = MP_OBJ_TO_PTR(arg0);
     struct addrinfo *res;
+    bool blocking = false;
+    bool connect_timeout = false;
+    int flags;
+    int raise_err = 0;
+    struct timeval timeout = { 0 };
+    socklen_t socklen;
+
     _socket_getaddrinfo(arg1, &res);
     MP_THREAD_GIL_EXIT();
     self->state = SOCKET_STATE_CONNECTED;
```

Comment:
I haven't changed this behaviour, but I noticed that this state transition happens even when `connect()` fails (meaning the check in `_socket_read_data` will pass through the read request where previously it would have failed it.)

I think a better thing to do here would be to delay changing state until the end of this function before it returns, or to change the state here and then change it back if raising an exception.

However, I'm worried there's some implicit race condition or something that I'm missing which is why it was set early to begin with...?

Reply:
You are right, this seems wrong to set the state here, instead of after the connection succeeds.

But, it would need some tests to see exactly what the expected (CPython) behaviour is when a connect fails.  And there are different failure modes: timeout, host unreachable, and maybe others.

See commit 7f366a2190825555c16f57f5dfd4d0d57efd7c1f which introduced this state change.  It has tests associated with it.

I think we should leave it as-is and tackle that issue separately, if it is indeed an issue.  First thing would be to write some tests for the case where a connect fails.